src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,162 @@
+      });
+
+    _kafka_has_sasl = ep_it != kafka_api.end();
+
+    _clean_timer.set_callback([this] {
+        // Do not need a gate since there are limited items
+        // in the cache. Therefore, relative to other design,
+        // garbage collection is inexpensive here.
+        clean_stale_clients();
+    });
+    _clean_timer.arm_periodic(clean_timer_period);
+}
+
+kafka_client_cache::~kafka_client_cache() { _clean_timer.cancel(); }
+
+client_ptr kafka_client_cache::fetch(credential_t user) {
+    auto& inner_hash = _cache.get<underlying_hash>();
+    auto it_hash = inner_hash.find(user.name);
+
+    if (it_hash == inner_hash.end()) {
+        return nullptr;
+    }
+
+    // Otherwise user-client found.
+
+    // Convert the hash iterator to list iterator
+    auto it_list = _cache.project<underlying_list>(it_hash);
+    auto& inner_list = _cache.get<underlying_list>();
+
+    // Extract the item. This is OK because elements are
+    // not copied or moved. Instead internal pointers
+    // are simply re-balanced and no allocations are made.
+    auto node = inner_list.extract(it_list);
+
+    // Update the timestamp and put it to the front
+    node.value().client->last_used = model::timestamp::now();
+    inner_list.push_front(node.value());
+
+    return node.value().client;
+}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_type authn_type) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when there is a Kafka listener with SASL
+    // enabled and the incoming request is using HTTP Basic AuthN
+    if (_kafka_has_sasl && authn_type == config::rest_authn_type::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(user.name);
+        cfg.scram_password.set_value(user.pass);
+    }
+
+    return ss::make_lw_shared<timestamped_client>(
+      to_yaml(cfg, config::redact_secrets::no), model::new_timestamp());
+}
+
+void kafka_client_cache::insert(credential_t user, client_ptr client) {
+    // Warn the user and do nothing if the
+    // max size is 0
+    if (_cache_max_size == 0) {
+        throw client_cache_error(
+          "Failed to insert client. Max cache size is 0.");
+    }
+
+    auto& inner_list = _cache.get<underlying_list>();
+
+    // First remove the last used client if the
+    // cache is full.
+    if (_cache.size() >= _cache_max_size) {
+        if (_cache.size() != 0) {
+            inner_list.pop_back();
+        }
+    }
+
+    user_client_pair pair{.username = user.name, .client = client};
+
+    // Add the user-client pair to front of frequency
+    // list since it will become recently "used"
+    // NOTE: push_front will silently fail if a duplicate
+    // already exists.
+    inner_list.push_front(pair);
```

Comment:
Should still use a gate/mutex here just-in-case multiple inserts for the same user occur concurrently.

This scenario will be covered in future scale tests where multiples users will issue rest requests.